README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![R-CMD-check](https://github.com/tshestakova/smosr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/tshestakova/smosr/actions/workflows/R-CMD-check.yaml)
[![CRAN
status](https://www.r-pkg.org/badges/version/smosr)](https://CRAN.R-project.org/package=smosr)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/smosr)](https://cran.r-project.org/package=smosr)
<!-- badges: end -->

# smosr

Acquire and Explore BEC-SMOS L4 Soil Moisture Data in R

## Overview

The goal of `smosr` is to automate accessing, downloading and exploring
Soil Moisture and Ocean Salinity (SMOS) Level 4 (L4) data developed by
Barcelona Expert Center (BEC). Particularly, it includes functions to
search for, acquire, extract, and plot BEC-SMOS L4 soil moisture data
downscaled to ~1 km spatial resolution.

## Installation

Install the released version of `smosr` from
[CRAN](https://cran.r-project.org/):

``` r
install.packages("smosr")
```

### Development version

To get a bug fix or to use the newest features of the package, install
the development version of `smosr` from [GitHub](https://github.com/):

``` r
# install.packages("devtools")
devtools::install_github("tshestakova/smosr")
```

## Usage

``` r
library(smosr)

# # to set credentials for accessing the BEC server
# # note that "username" and "password" should be replaced with your login details 
# set_credentials("username", "password")

# to search for BEC-SMOS data with the specified parameters available on the server
start_date <- as.Date("2010-07-15")
end_date <- as.Date("2022-07-15")
date_range <- seq(start_date, end_date, by = 365)
smos_data <- find_smos(freq = 3, orbit = "des", dates = date_range)
#> Done. All requested files were successfully found.

# to download the data from the server to a temporary directory in a local computer
download_smos(data = smos_data)
#> |================================================================================| 100%

# to list all BEC-SMOS data files currently stored in the temporary directory
smos_files <- list_smos()
head(smos_files, 5)
#> [1] "C:\\Users\\...\\Temp\\RtmpojyuH6/BEC_SM____SMOS__EUM_L4__D_20100715T183618_001km_3d_REP_v6.0.nc"
#> [2] "C:\\Users\\...\\Temp\\RtmpojyuH6/BEC_SM____SMOS__EUM_L4__D_20110715T182802_001km_3d_REP_v6.0.nc"
#> [3] "C:\\Users\\...\\Temp\\RtmpojyuH6/BEC_SM____SMOS__EUM_L4__D_20120714T182030_001km_3d_REP_v6.0.nc"
#> [4] "C:\\Users\\...\\Temp\\RtmpojyuH6/BEC_SM____SMOS__EUM_L4__D_20130714T181231_001km_3d_REP_v6.0.nc"
#> [5] "C:\\Users\\...\\Temp\\RtmpojyuH6/BEC_SM____SMOS__EUM_L4__D_20140714T180441_001km_3d_REP_v6.0.nc"

# to extract soil moisture estimates for the specified geographical locations
# from the list of data files obtained in the previous steps
lat <- c(38.72, 41.90, 48.86, 52.50, 59.44)
lon <- c(-9.14, 12.50, 2.35, 13.40, 24.75)
sm_estimates <- extract_smos(data = smos_files, lat = lat, lon = lon)
#> |================================================================================| 100%

head(sm_estimates, 12)
#>       Lat     Lon     Freq Orbit Date         Time       SM                     QA
#>  [1,] "38.72" "-9.14" "3d" "D"   "2010-07-15" "20:36:18" "0.0176999995528604"   "5"
#>  [2,] "41.9"  "12.5"  "3d" "D"   "2010-07-15" "20:36:18" "0.0872999977946165"   "7"
#>  [3,] "48.86" "2.35"  "3d" "D"   "2010-07-15" "20:36:18" "0.206299994788424"    "2"
#>  [4,] "52.5"  "13.4"  "3d" "D"   "2010-07-15" "20:36:18" "0.191299995167356"    "6"
#>  [5,] "59.44" "24.75" "3d" "D"   "2010-07-15" "20:36:18" "0.165999995806487"    "7"
#>  [6,] "38.72" "-9.14" "3d" "D"   "2011-07-15" "20:28:02" "0.071999998181127"    "3"
#>  [7,] "41.9"  "12.5"  "3d" "D"   "2011-07-15" "20:28:02" "0.025399999358342"    "7"
#>  [8,] "48.86" "2.35"  "3d" "D"   "2011-07-15" "20:28:02" "0.0235999994038139"   "6"
#>  [9,] "52.5"  "13.4"  "3d" "D"   "2011-07-15" "20:28:02" "0.0769999980548164"   "2"
#> [10,] "59.44" "24.75" "3d" "D"   "2011-07-15" "20:28:02" "0.144699996344571"    "7"
#> [11,] "38.72" "-9.14" "3d" "D"   "2012-07-14" "20:20:31" "0.0246999993760255"   "5"
#> [12,] "41.9"  "12.5"  "3d" "D"   "2012-07-14" "20:20:31" "0.000499999987368938" "7"

# to plot extracted temporal series of soil moisture data
plot_temporal_smos(data = sm_estimates)
```

<img src="man/figures/README-example-temp.png"/>

``` r
# to draw a raster image of soil moisture estimates from a file with
# desired data quality and within the specified geographical bounds
lat <- c(35.00, 45.00)
lon <- c(-10.50, 4.50)
plot_raster_smos(data = smos_files[13], lat = lat, lon = lon, QA = c(0,1,2,3))
```

<img src="man/figures/README-example-sp.png"/>

## Getting help

If you encounter a clear bug, please file an issue with a minimal
reproducible example on
[GitHub](https://github.com/tshestakova/smosr/issues).

For questions and other discussion, please use [Stack
Overflow](https://stackoverflow.com/questions/) and the [RStudio
community](https://community.rstudio.com/).


NEWS.md
# smosr 1.0.1

* released to CRAN
* Added a `NEWS.md` file to track changes to the package

# smosr 1.0.0

* Initial CRAN submission


smosr.Rcheck/smosr-Ex.R
pkgname <- "smosr"
source(file.path(R.home("share"), "R", "examples-header.R"))
options(warn = 1)
options(pager = "console")
base::assign(".ExTimings", "smosr-Ex.timings", pos = 'CheckExEnv')
base::cat("name\tuser\tsystem\telapsed\n", file=base::get(".ExTimings", pos = 'CheckExEnv'))
base::assign(".format_ptime",
function(x) {
  if(!is.na(x[4L])) x[1L] <- x[1L] + x[4L]
  if(!is.na(x[5L])) x[2L] <- x[2L] + x[5L]
  options(OutDec = '.')
  format(x[1L:3L], digits = 7L)
},
pos = 'CheckExEnv')

### * </HEADER>
library('smosr')

base::assign(".oldSearch", base::search(), pos = 'CheckExEnv')
base::assign(".old_wd", base::getwd(), pos = 'CheckExEnv')
cleanEx()
nameEx("download_smos")
### * download_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: download_smos
### Title: Download BEC-SMOS soil moisture data
### Aliases: download_smos

### ** Examples

## Not run: 
##D # to download all files found with find_smos()
##D # into the current working directory
##D start_date <- as.Date("2022-01-01")
##D end_date <- as.Date("2022-12-31")
##D date_range <- seq(start_date, end_date, by = 30)
##D smos_data <- find_smos(freq = 3, orbit = "descending", dates = date_range)
##D download_smos(smos_data)
##D # to download first five items from the complete list of files
##D # and place them in a newly created directory
##D dir.create("~/SMOS_data")
##D download_smos(data = smos_data[1:5], dir = "~/SMOS_data")
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("download_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("extract_smos")
### * extract_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: extract_smos
### Title: Extract BEC-SMOS soil moisture estimates for specific
###   geographical locations
### Aliases: extract_smos

### ** Examples

## Not run: 
##D # to iterate over BEC-SMOS data files stored in the current working directory
##D # and extract soil moisture estimates for the specified geographical locations
##D smos_data <- list_smos()
##D lat <- c(40.42, 41.90, 48.86, 52.50, 59.91)
##D lon <- c(-3.70, 12.50, 2.35, 13.40, 10.75)
##D sm_estimates <- extract_smos(data = smos_data, lat = lat, lon = lon)
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("extract_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("find_smos")
### * find_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: find_smos
### Title: Find BEC-SMOS soil moisture data in Barcelona Expert Center
###   (BEC) server
### Aliases: find_smos

### ** Examples

## Not run: 
##D # to look for SMOS data on a specific date
##D smos_data <- find_smos(freq = 1, orbit = "a", dates = "2022-12-31")
##D # to search over a date range
##D start_date <- as.Date("2022-01-01")
##D end_date <- as.Date("2022-12-31")
##D date_range <- seq(start_date, end_date, by = 10)
##D smos_data <- find_smos(freq = 3, orbit = "descending", dates = date_range)
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("find_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("list_smos")
### * list_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: list_smos
### Title: List the BEC-SMOS data files stored on a local computer
### Aliases: list_smos

### ** Examples

## Not run: 
##D # to list all BEC-SMOS data files stored in the current working directory
##D # as well as in the corresponding subfolders
##D smos_data <- list_smos(recursive = TRUE)
##D # to list BEC-SMOS data files with the specified frequency and SMOS orbit
##D # stored in the specified folder
##D smos_data <- list_smos(freq = 3, orbit = "asc", dir = "~/SMOS_data")
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("list_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("missing_smos")
### * missing_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: missing_smos
### Title: Print the dates for which BEC-SMOS soil moisture data have not
###   been found
### Aliases: missing_smos

### ** Examples

## Not run: 
##D missing_smos()
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("missing_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("plot_raster_smos")
### * plot_raster_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: plot_raster_smos
### Title: Draw a raster image of BEC-SMOS soil moisture data
### Aliases: plot_raster_smos

### ** Examples

## Not run: 
##D # to draw a raster image of soil moisture data within specified bounds
##D smos_data <- list_smos()
##D lat <- c(35.00, 45.00)
##D lon <- c(-10.50, 4.50)
##D plot_raster_smos(data = smos_data[1], lat = lat, lon = lon)
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("plot_raster_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("plot_temporal_smos")
### * plot_temporal_smos

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: plot_temporal_smos
### Title: Plot temporal series of BEC-SMOS soil moisture data
### Aliases: plot_temporal_smos

### ** Examples

## Not run: 
##D # to plot extracted temporal series of soil moisture data
##D # with the specified frequency, SMOS orbit and QA
##D smos_data <- list_smos()
##D lat <- c(40.42, 41.90, 48.86, 52.50, 59.91)
##D lon <- c(-3.70, 12.50, 2.35, 13.40, 10.75)
##D sm_estimates <- extract_smos(data = smos_data, lat = lat, lon = lon)
##D plot_temporal_smos(data = sm_estimates, freq = 3, orbit = "d", QA = 0)
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("plot_temporal_smos", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
cleanEx()
nameEx("set_credentials")
### * set_credentials

flush(stderr()); flush(stdout())

base::assign(".ptime", proc.time(), pos = "CheckExEnv")
### Name: set_credentials
### Title: Set credentials to access Barcelona Expert Center (BEC) server
### Aliases: set_credentials

### ** Examples

## Not run: 
##D set_credentials("username", "password")
## End(Not run)




base::assign(".dptime", (proc.time() - get(".ptime", pos = "CheckExEnv")), pos = "CheckExEnv")
base::cat("set_credentials", base::get(".format_ptime", pos = 'CheckExEnv')(get(".dptime", pos = "CheckExEnv")), "\n", file=base::get(".ExTimings", pos = 'CheckExEnv'), append=TRUE, sep="\t")
### * <FOOTER>
###
cleanEx()
options(digits = 7L)
base::cat("Time elapsed: ", proc.time() - base::get("ptime", pos = 'CheckExEnv'),"\n")
grDevices::dev.off()
###
### Local variables: ***
### mode: outline-minor ***
### outline-regexp: "\\(> \\)?### [*]+" ***
### End: ***
quit('no')
